euphonic/spectra.py

```diff
@@ -1297,6 +1298,119 @@ def from_dict(cls: Type[T], d: Dict[str, Any]) -> T:
                    x_tick_labels=d['x_tick_labels'],
                    metadata=d['metadata'])
 
+def apply_kinematic_constraints(spectrum: Spectrum2D,
+                                e_i: Quantity = None,
+                                e_f: Quantity = None,
+                                angle_range: Tuple[float] = (0, 180.)
+                                ) -> Spectrum2D:
+    """Set events to NaN which violate energy/momentum limits:
+
+      - Energy transfer greater than e_i
+      - q outside region accessible for given e_i and angle range
+
+    This requires x_data to be in wavevector units and y_data to be energy.
+
+    Either e_i or e_f should be set, according to direct/instrument
+    geometry. The other values will be inferred, interpreting y_data as
+    energy transfer.
+
+    Args:
+        spectrum: input 2-D spectrum, with |q| x_data and energy y_data
+        e_i: incident energy of direct-geometry spectrometer
+        e_f: final energy of indirect-geometry spectrometer
+        angle_range: min and max scattering angles (2θ) of detector bank in
+            degrees.
+
+    Returns:
+        Masked spectrum with inaccessible bins set to NaN in z_data.
+    """
```

Comment:
This should be a Numpy-style docstring like other docstrings in Euphonic. And the """ should be on a line on its own with the text starting on a new line.